内核与编译路径不同时驱动编译错误

### Xmake 版本

v2.8.5+master

### 操作系统版本和架构

ubuntu23.10

### 描述问题

linux 内核正常情况下编译是直接进去目录 make config -> make 这样编译路径和代码路径是相同的
很多sdk给的编译脚本内核路径和编译路径是不一样的
    我们假设内核代码路径是A ， 编译路径是 B  
    那他们配置编译使用
        make -C ${A} O=${B} a_config 
        make -C ${A} O=${B}
    这样的编译命令

这样的话编译驱动模块的时候 
    make -C ${A} M=${pwd} module时候会报`Kernel configuration is invalid.` 因为在A这个目录下根本没有配置文件
需要加上 -O ${B} 参数带上内核的配置文件的路径
    make -C ${A} O=${B} M=${pwd} module 这样的编译命令才能正确编译模块

### 期待的结果

希望xmake也能支持这样的模块编译 同样的 需要加入外部传入kernel编译路径

### 工程配置

ifneq ($(KERNELRELEASE),)
    obj-m := hello.o
	hello-objs := ./src/hello.o ./src/add.o
else
    KERN_DIR ?= /home/wangyang/luckfox-pico/sysdrv/source/kernel
    PWD := $(shell pwd)

default:
	$(MAKE) ARCH=arm CROSS_COMPILE=/home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf- -C $(KERN_DIR) V=1 M=$(PWD) O=/home/wangyang/luckfox-pico/sysdrv/source/objs_kernel modules
endif

clean:
	rm -rf *.o *~ core .depend .*.cmd *.ko *.mod.c .tmp_versions
	rm -rf src/*.o src/*~ core src/.depend src/.*.cmd src/*.ko src/*.mod.c src/.tmp_versions


### 附加信息和错误日志

tests/projects/linux/driver/hello_makefile/Makefile模块编译日志 上面是没改的 下面改过参数 -O ${buildpath}
wangyang@wangyang:~/xmake/tests/projects/linux/driver/hello_makefile$ make V=1
make ARCH=arm CROSS_COMPILE=/home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf- -C /home/wangyang/luckfox-pico/sysdrv/source/kernel V=1 M=/home/wangyang/xmake/tests/projects/linux/driver/hello_makefile modules
make[1]: Entering directory '/home/wangyang/luckfox-pico/sysdrv/source/kernel'
test -e include/generated/autoconf.h -a -e include/config/auto.conf || (                \
echo >&2;                                                       \
echo >&2 "  ERROR: Kernel configuration is invalid.";           \
echo >&2 "         include/generated/autoconf.h or include/config/auto.conf are missing.";\
echo >&2 "         Run 'make oldconfig && make prepare' on kernel src to fix it.";      \
echo >&2 ;                                                      \
/bin/false)

  ERROR: Kernel configuration is invalid.
         include/generated/autoconf.h or include/config/auto.conf are missing.
         Run 'make oldconfig && make prepare' on kernel src to fix it.

Makefile:672: include/config/auto.conf: No such file or directory
make[1]: *** [Makefile:758: include/config/auto.conf] Error 1
make[1]: Leaving directory '/home/wangyang/luckfox-pico/sysdrv/source/kernel'
make: *** [Makefile:9: default] Error 2
wangyang@wangyang:~/xmake/tests/projects/linux/driver/hello_makefile$ 
wangyang@wangyang:~/xmake/tests/projects/linux/driver/hello_makefile$ 
wangyang@wangyang:~/xmake/tests/projects/linux/driver/hello_makefile$ make V=1
make ARCH=arm CROSS_COMPILE=/home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf- -C /home/wangyang/luckfox-pico/sysdrv/source/kernel V=1 M=/home/wangyang/xmake/tests/projects/linux/driver/hello_makefile O=/home/wangyang/luckfox-pico/sysdrv/source/objs_kernel modules
make[1]: Entering directory '/home/wangyang/luckfox-pico/sysdrv/source/kernel'
make -C /home/wangyang/luckfox-pico/sysdrv/source/objs_kernel -f /home/wangyang/luckfox-pico/sysdrv/source/kernel/Makefile modules
make[2]: Entering directory '/home/wangyang/luckfox-pico/sysdrv/source/objs_kernel'
test -e include/generated/autoconf.h -a -e include/config/auto.conf || (                \
echo >&2;                                                       \
echo >&2 "  ERROR: Kernel configuration is invalid.";           \
echo >&2 "         include/generated/autoconf.h or include/config/auto.conf are missing.";\
echo >&2 "         Run 'make oldconfig && make prepare' on kernel src to fix it.";      \
echo >&2 ;                                                      \
/bin/false)
make -f /home/wangyang/luckfox-pico/sysdrv/source/kernel/scripts/Makefile.build obj=/home/wangyang/xmake/tests/projects/linux/driver/hello_makefile \
single-build= \
need-builtin=1 need-modorder=1
  /home/wangyang/luckfox-pico/sysdrv/source/kernel/scripts/gcc-wrapper.py /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf-gcc -Wp,-MMD,/home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/.hello.o.d -nostdinc -isystem /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/../lib/gcc/arm-rockchip830-linux-uclibcgnueabihf/8.3.0/include -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/arch/arm/include -I./arch/arm/include/generated -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/include -I./include -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/arch/arm/include/uapi -I./arch/arm/include/generated/uapi -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/include/uapi -I./include/generated/uapi -include /home/wangyang/luckfox-pico/sysdrv/source/kernel/include/linux/kconfig.h -include /home/wangyang/luckfox-pico/sysdrv/source/kernel/include/linux/compiler_types.h -D__KERNEL__ -mlittle-endian -fmacro-prefix-map=/home/wangyang/luckfox-pico/sysdrv/source/kernel/= -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security -std=gnu89 -fno-dwarf2-cfi-asm -fno-ipa-sra -mabi=aapcs-linux -mfpu=vfp -funwind-tables -mthumb -Wa,-mimplicit-it=always -Wa,-mno-warn-deprecated -D__LINUX_ARM_ARCH__=7 -march=armv7-a -msoft-float -Uarm -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Os --param=allow-store-data-races=0 -Wframe-larger-than=1024 -fstack-protector -Werror -Wimplicit-fallthrough -Wno-unused-but-set-variable -Wno-unused-const-variable -fomit-frame-pointer -g -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Wno-stringop-truncation -Wno-array-bounds -Wno-stringop-overflow -Wno-restrict -Wno-maybe-uninitialized -fno-strict-overflow -fno-stack-check -fconserve-stack -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init -Wno-packed-not-aligned  -DMODULE  -DKBUILD_BASENAME='"hello"' -DKBUILD_MODNAME='"hello"' -D__KBUILD_MODNAME=kmod_hello -c -o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/hello.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/hello.c
  /home/wangyang/luckfox-pico/sysdrv/source/kernel/scripts/gcc-wrapper.py /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf-gcc -Wp,-MMD,/home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/.add.o.d -nostdinc -isystem /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/../lib/gcc/arm-rockchip830-linux-uclibcgnueabihf/8.3.0/include -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/arch/arm/include -I./arch/arm/include/generated -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/include -I./include -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/arch/arm/include/uapi -I./arch/arm/include/generated/uapi -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/include/uapi -I./include/generated/uapi -include /home/wangyang/luckfox-pico/sysdrv/source/kernel/include/linux/kconfig.h -include /home/wangyang/luckfox-pico/sysdrv/source/kernel/include/linux/compiler_types.h -D__KERNEL__ -mlittle-endian -fmacro-prefix-map=/home/wangyang/luckfox-pico/sysdrv/source/kernel/= -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security -std=gnu89 -fno-dwarf2-cfi-asm -fno-ipa-sra -mabi=aapcs-linux -mfpu=vfp -funwind-tables -mthumb -Wa,-mimplicit-it=always -Wa,-mno-warn-deprecated -D__LINUX_ARM_ARCH__=7 -march=armv7-a -msoft-float -Uarm -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Os --param=allow-store-data-races=0 -Wframe-larger-than=1024 -fstack-protector -Werror -Wimplicit-fallthrough -Wno-unused-but-set-variable -Wno-unused-const-variable -fomit-frame-pointer -g -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Wno-stringop-truncation -Wno-array-bounds -Wno-stringop-overflow -Wno-restrict -Wno-maybe-uninitialized -fno-strict-overflow -fno-stack-check -fconserve-stack -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init -Wno-packed-not-aligned  -DMODULE  -DKBUILD_BASENAME='"add"' -DKBUILD_MODNAME='"hello"' -D__KBUILD_MODNAME=kmod_hello -c -o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/add.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/add.c
  /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf-ld -EL -z noexecstack   -r -o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/hello.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/add.o
  { echo  /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/hello.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/./src/add.o;  echo; } > /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.mod
  {   echo /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.ko; :; } | awk '!x[$0]++' - > /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/modules.order
make -f /home/wangyang/luckfox-pico/sysdrv/source/kernel/scripts/Makefile.modpost
  sed 's/\.ko$/\.o/' /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/modules.order | scripts/mod/modpost     -o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/Module.symvers -e -i Module.symvers   -T -
make -f /home/wangyang/luckfox-pico/sysdrv/source/kernel/scripts/Makefile.modfinal
  /home/wangyang/luckfox-pico/sysdrv/source/kernel/scripts/gcc-wrapper.py /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf-gcc -Wp,-MMD,/home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/.hello.mod.o.d -nostdinc -isystem /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/../lib/gcc/arm-rockchip830-linux-uclibcgnueabihf/8.3.0/include -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/arch/arm/include -I./arch/arm/include/generated -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/include -I./include -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/arch/arm/include/uapi -I./arch/arm/include/generated/uapi -I/home/wangyang/luckfox-pico/sysdrv/source/kernel/include/uapi -I./include/generated/uapi -include /home/wangyang/luckfox-pico/sysdrv/source/kernel/include/linux/kconfig.h -include /home/wangyang/luckfox-pico/sysdrv/source/kernel/include/linux/compiler_types.h -D__KERNEL__ -mlittle-endian -fmacro-prefix-map=/home/wangyang/luckfox-pico/sysdrv/source/kernel/= -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security -std=gnu89 -fno-dwarf2-cfi-asm -fno-ipa-sra -mabi=aapcs-linux -mfpu=vfp -funwind-tables -mthumb -Wa,-mimplicit-it=always -Wa,-mno-warn-deprecated -D__LINUX_ARM_ARCH__=7 -march=armv7-a -msoft-float -Uarm -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Os --param=allow-store-data-races=0 -Wframe-larger-than=1024 -fstack-protector -Werror -Wimplicit-fallthrough -Wno-unused-but-set-variable -Wno-unused-const-variable -fomit-frame-pointer -g -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Wno-stringop-truncation -Wno-array-bounds -Wno-stringop-overflow -Wno-restrict -Wno-maybe-uninitialized -fno-strict-overflow -fno-stack-check -fconserve-stack -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init -Wno-packed-not-aligned -DMODULE -DKBUILD_BASENAME='"hello.mod"' -DKBUILD_MODNAME='"hello"' -D__KBUILD_MODNAME=kmod_hello -c -o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.mod.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.mod.c
  /home/wangyang/luckfox-pico/tools/linux/toolchain/arm-rockchip830-linux-uclibcgnueabihf/bin/arm-rockchip830-linux-uclibcgnueabihf-ld -r -EL -z noexecstack --build-id=sha1  -T scripts/module.lds -o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.ko /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.o /home/wangyang/xmake/tests/projects/linux/driver/hello_makefile/hello.mod.o;  true
make[2]: Leaving directory '/home/wangyang/luckfox-pico/sysdrv/source/objs_kernel'
make[1]: Leaving directory '/home/wangyang/luckfox-pico/sysdrv/source/kernel

--- a/xmake/rules/platform/linux/driver/driver_modules.lua
+++ b/xmake/rules/platform/linux/driver/driver_modules.lua
@@ -31,15 +31,11 @@ function _get_linux_headers_sdk(target)
     local linux_headersdir = target:values("linux.driver.linux-headers")
     local linux_builddir = target:values("linux.driver.linux-builddir")
     if linux_headersdir then
-        if linux_builddir then
-            return {
-                sdkdir = linux_headersdir,
-                builddir = linux_builddir,
-                includedir = path.join(linux_headersdir, "include")
-            }
-        else
-            return {sdkdir = linux_headersdir, includedir = path.join(linux_headersdir, "include")}
-        end
+        return {
+            sdkdir = linux_headersdir,
+            builddir = linux_builddir,
+            includedir = path.join(linux_headersdir, "include")
+        }
     end
     local linux_headers = assert(target:pkg("linux-headers"), "please add `add_requires(\"linux-headers\", {configs = {driver_modules = true}})` and `add_packages(\"linux-headers\")` to the given target!")
     local includedirs = linux_headers:get("includedirs") or linux_headers:get("sysincludedirs")
@@ -151,11 +147,7 @@ module_exit(hello_exit);
                                 includedir = cflag:sub(3)
                             end
                             if not path.is_absolute(includedir) then
-                                if builddir then
-                                    includedir = path.absolute(includedir, builddir)
-                                else
-                                    includedir = path.absolute(includedir, sdkdir)
-                                end
+                                includedir = path.absolute(includedir, builddir or sdkdir)
                             end
                             if cflag:startswith("-I") then
                                 cflag = "-I" .. includedir
@@ -177,11 +169,7 @@ module_exit(hello_exit);
                     for _, ldflag in ipairs(os.argv(ldflags)) do
                         if ldflag:endswith(".lds") then
                             if not path.is_absolute(ldflag) then
-                                if builddir then
-                                    ldflag = path.absolute(ldflag, builddir)
-                                else
-                                    ldflag = path.absolute(ldflag, sdkdir)
-                                end
+                                ldflag = path.absolute(ldflag, builddir or sdkdir)
                             end
                         end
                         if ko then
@@ -265,11 +253,7 @@ function link(target, opt)
         local modpost
         local linux_headers = target:data("linux.driver.linux_headers")
         if linux_headers then
-            if linux_headers.builddir then
-                modpost = path.join(linux_headers.builddir, "scripts", "mod", "modpost")
-            else
-                modpost = path.join(linux_headers.sdkdir, "scripts", "mod", "modpost")
-            end
+            modpost = path.join(linux_headers.builddir or linux_headers.sdkdir, "scripts", "mod", "modpost")
         end
         assert(modpost and os.isfile(modpost), "scripts/mod/modpost not found!")
 

--- a/xmake/rules/platform/linux/driver/driver_modules.lua
+++ b/xmake/rules/platform/linux/driver/driver_modules.lua
@@ -29,8 +29,17 @@ import("private.tools.ccache")
 -- get linux-headers sdk
 function _get_linux_headers_sdk(target)
     local linux_headersdir = target:values("linux.driver.linux-headers")
+    local linux_builddir = target:values("linux.driver.linux-builddir")
     if linux_headersdir then
-        return {sdkdir = linux_headersdir, includedir = path.join(linux_headersdir, "include")}
+        if linux_builddir then
+            return {
+                sdkdir = linux_headersdir,
+                builddir = linux_builddir,
+                includedir = path.join(linux_headersdir, "include")
+            }
+        else
+            return {sdkdir = linux_headersdir, includedir = path.join(linux_headersdir, "include")}
+        end
     end
     local linux_headers = assert(target:pkg("linux-headers"), "please add `add_requires(\"linux-headers\", {configs = {driver_modules = true}})` and `add_packages(\"linux-headers\")` to the given target!")
     local includedirs = linux_headers:get("includedirs") or linux_headers:get("sysincludedirs")
@@ -52,7 +61,7 @@ function _get_linux_headers_sdk(target)
 end
 
 -- get cflags from make
-function _get_cflags_from_make(target, sdkdir)
+function _get_cflags_from_make(target, sdkdir, builddir)
     local key = sdkdir .. target:arch()
     local cflags = memcache.get2("linux.driver", key, "cflags")
     local ldflags_o = memcache.get2("linux.driver", key, "ldflags_o")
@@ -88,6 +97,9 @@ module_init(hello_init);
 module_exit(hello_exit);
         ]])
         local argv = {"-C", sdkdir, "V=1", "M=" .. tmpdir, "modules"}
+        if builddir then
+            table.insert(argv, "O=" .. builddir)
+        end
         if not target:is_plat(os.subhost()) then
             -- e.g.	$(MAKE) -C $(KERN_DIR) V=1 ARCH=arm64 CROSS_COMPILE=/mnt/gcc-linaro-7.5.0-2019.12-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu- M=$(PWD) modules
             local arch
@@ -139,7 +151,11 @@ module_exit(hello_exit);
                                 includedir = cflag:sub(3)
                             end
                             if not path.is_absolute(includedir) then
-                                includedir = path.absolute(includedir, sdkdir)
+                                if builddir then
+                                    includedir = path.absolute(includedir, builddir)
+                                else
+                                    includedir = path.absolute(includedir, sdkdir)
+                                end
                             end
                             if cflag:startswith("-I") then
                                 cflag = "-I" .. includedir
@@ -161,7 +177,11 @@ module_exit(hello_exit);
                     for _, ldflag in ipairs(os.argv(ldflags)) do
                         if ldflag:endswith(".lds") then
                             if not path.is_absolute(ldflag) then
-                                ldflag = path.absolute(ldflag, sdkdir)
+                                if builddir then
+                                    ldflag = path.absolute(ldflag, builddir)
+                                else
+                                    ldflag = path.absolute(ldflag, sdkdir)
+                                end
                             end
                         end
                         if ko then
@@ -224,7 +244,7 @@ function config(target)
     for _, sourcefile in ipairs(target:sourcefiles()) do
         target:fileconfig_set(sourcefile, {defines = "KBUILD_BASENAME=\"" .. path.basename(sourcefile) .. "\""})
     end
-    local cflags, ldflags_o, ldflags_ko = _get_cflags_from_make(target, linux_headers.sdkdir)
+    local cflags, ldflags_o, ldflags_ko = _get_cflags_from_make(target, linux_headers.sdkdir, linux_headers.builddir)
     if cflags then
         target:add("cflags", cflags, {force = true})
         target:data_set("linux.driver.ldflags_o", ldflags_o)
@@ -245,7 +265,11 @@ function link(target, opt)
         local modpost
         local linux_headers = target:data("linux.driver.linux_headers")
         if linux_headers then
-            modpost = path.join(linux_headers.sdkdir, "scripts", "mod", "modpost")
+            if linux_headers.builddir then
+                modpost = path.join(linux_headers.builddir, "scripts", "mod", "modpost")
+            else
+                modpost = path.join(linux_headers.sdkdir, "scripts", "mod", "modpost")
+            end
         end
         assert(modpost and os.isfile(modpost), "scripts/mod/modpost not found!")
 

